### PR TITLE
RDoc-2482 Python demos (Basics + Related documents)

### PR DESCRIPTION
**Pull request by Danielle9897**
**Related issue:**
https://issues.hibernatingrhinos.com/issue/RDoc-2482/Python-Demo-examples

---

@poissoncorp This PR has the following demos:

**Basics category:**
The document store
The session
Create document
Load and edit document
Delete document

**Related documents category:**
Create related documents
Load related documents
Query related documents
Index related documents

### CODE REVIEW

File: DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            "Title": "Create a new Python object to be saved",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* Create a Python object to be saved.",
+                "",
+                "* By default, _the Python class name_ determines the collection to which the document will belong to in the database.",
+                "  Collections are used to group documents together so that it is convenient to apply some operation to them.",
+                "",
+                "* _The Python class members_ determine the document JSON structure in the database.<br>",
+                "  The JSON serialization for the document fields follows the conventions set on the Document Store",
```
Comment (poissoncorp) [inline DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json:37]: In our Python client, JSON serialization process is customizable, but don't use mappers/conventions. Let's talk directly about this

File: DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            "Title": "The new document ID is immediately available",
+            "Slug" : "step-3",
+            "DescriptionLines": [
+                "* The Session takes care of generating the document identifier automatically.<br>",
+                "  The new stored entity ID is now available and can be used.",
+                "",
+                "* _If your Python class was defined with an 'Id' property_:<br>",
+                "  => You can get the new stored entity ID as shown in the above example.",
+                "",
+                "* _If your Python class was Not defined with an 'Id' property_:<br>",
```
Comment (poissoncorp) [inline DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json:87]: even when we don't specify 'Id' field in the class definition, the client will add such a field to the object, taking leverage of Python dynamic typing

File: DemoServer/AdditionalLanguages/python/basics/deleteDocument/metadata.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            "Title": "Mark the entity to be deleted",
+            "Slug" : "step-1",
+            "DescriptionLines": [
+                "* Mark the entity to be deleted with the Session `delete()` method.<br>",
+                "  The document is actually deleted from the database only when save_changes() is called later on.",
+                "",
+                "_Note_:",
+                "",
+                "* When using delete() as above, with the specific ID param overload, the document is deleted with no concurrency checks.",
```
Comment (poissoncorp) [inline DemoServer/AdditionalLanguages/python/basics/deleteDocument/metadata.json:21]: we should avoid the 'overload' word, to be pure

in Python, there are no overloads - we either check the argument type passed to the method or define two different methods with different names like `delete_by_id` or `delete_by_entity `

in our case, `delete()` takes `key_or_entity` argument

```py
def delete(self, key_or_entity: Union[str, object], expected_change_vector: Optional[str] = None) -> None:
        if isinstance(key_or_entity, str):
```

File: DemoServer/AdditionalLanguages/python/basics/theDocumentStore/metadata.json [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            "Slug" : "step-4",
+            "DescriptionLines": [
+                "* At this point, before initializing the Document Store (in step 5), the following info can be configured:",
+                "",
+                "  * _urls_ - Specify a list of URL addresses pointing to the RavenDB server nodes composing your cluster.",
+                "    Note: Do not open a DocumentStore using URL addresses that point to nodes outside your cluster.",
+                "",
+                "  * _database_ - Specify the default database name this Document Store will work against.",
+                "",
+                "  * _certificate_ - Specify the Client Certificate ( *.pfx ) which is used to authenticate the client to the RavenDB server.",
```
Comment (poissoncorp) [inline DemoServer/AdditionalLanguages/python/basics/theDocumentStore/metadata.json:81]:
```suggestion
                "  * _certificate_ - Specify the Client Certificate ( *.pem) which is used to authenticate the client to the RavenDB server.",
```

**Review by poissoncorp — CHANGES_REQUESTED** [review]
The first issue with conventions is definitely thing to do, but "overload" is a minor one :)
Overall, looks very good to me

File: DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json (hunk 2 above)
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/python/basics/createDocument/metadata.json:87]: done

File: DemoServer/AdditionalLanguages/python/basics/deleteDocument/metadata.json (hunk 3 above)
Comment (Danielle9897) [reply DemoServer/AdditionalLanguages/python/basics/deleteDocument/metadata.json:21]: done

File: DemoServer/AdditionalLanguages/python/relatedDocuments/indexRelatedDocuments/indexRelatedDocuments.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+from models import Product
+from typing import List
+
+class RunParams(RunParamsBase):
+    def __init__(self, category_name: str):
+        self.category_name = category_name
+
+#region Demo
+#region Step_1
+class Products_ByCategoryName(AbstractIndexCreationTask):
```
Comment (Danielle9897) [inline DemoServer/AdditionalLanguages/python/relatedDocuments/indexRelatedDocuments/indexRelatedDocuments.py:15]: @poissoncorp 

pls let's talk about the following:

1. I removed the original enclosing class here
2. I think we need an inner 'IndexEntry' class - similar to other demos

**Review by poissoncorp — APPROVED** [review]
lets quickly talk about the IndexEntry

**Pull request merged**